### PR TITLE
FAI-12269: Bitbucket source base work

### PR DESCRIPTION
**Pull request by matiaslcoulougian**
## Description
Refactors on current source to make it consistent with our other connectors, and to remove duplication.

Changes:
- Created and implemented common stream with slices by workspace and repo.
- Refactored configs to make them consistent with other sources
- Reduced duplication - common types

TODO: 
- [x] Update missing streams to use common stream
- [x] Fix tests

## Type of change
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change

## Related issues

> Fix [#1]() 

## Migration notes

> Describe migration notes if any

## Extra info

> Add any additional information

### CODE REVIEW

File: faros-airbyte-common/src/common/index.ts [hunk 1]
```diff
@@ -74,3 +74,21 @@ export function calculateDateRange(options: {
 
   return {startDate, endDate};
 }
+
+export function collectReposByNamespace(
```
Comment (ypc-faros) [inline faros-airbyte-common/src/common/index.ts:78]: Why not `collectReposByOrg` ?

Comment (ypc-faros) [inline faros-airbyte-common/src/common/index.ts:78]: Let's add a couple of tests for this

Comment (matiaslcoulougian) [reply faros-airbyte-common/src/common/index.ts:78]: I thought of a more generic name, because for Bitbucket they is called workspace

Comment (matiaslcoulougian) [reply faros-airbyte-common/src/common/index.ts]: I changed it back to collectReposByOrg

File: sources/bitbucket-source/src/bitbucket.ts [hunk 2]
```diff
@@ -60,14 +62,12 @@ export class Bitbucket {
 
     const baseUrl = config.serverUrl || DEFAULT_BITBUCKET_URL;
     const client = new BitbucketClient({baseUrl, auth});
-    const pagelen = config.pagelen || DEFAULT_PAGELEN;
+    const pageSize = config.page_size || DEFAULT_PAGE_SIZE;
 
-    if (!config.cutoff_days) {
-      throw new VError('cutoff_days is null or empty');
-    }
+    const cutoffDays = config.cutoff_days ?? DEFAULT_CUTOFF_DAYS;
```
Comment (ypc-faros) [inline sources/bitbucket-source/src/bitbucket.ts:67]: Let's do the same as in Jira, GitHub: https://github.com/faros-ai/airbyte-connectors/blob/9a0d63dbc7f81b9833607701f25058434b30546f/sources/jira-source/src/index.ts#L106-L111

Okay to do in a separate PR

File: sources/bitbucket-source/resources/spec.json [hunk 3]
```diff
@@ -4,10 +4,10 @@
     "$schema": "http://json-schema.org/draft-07/schema#",
     "title": "Bitbucket Spec",
     "type": "object",
-    "required": ["workspaces", "cutoff_days"],
+    "required": ["workspaces"],
     "additionalProperties": true,
     "properties": {
-      "serverUrl": {
+      "api_url": {
```
Comment (ypc-faros) [inline sources/bitbucket-source/resources/spec.json:10]: Since you're here, please add `order` to the props

File: sources/bitbucket-source/src/bitbucket.ts [hunk 4]
```diff
@@ -60,14 +62,12 @@ export class Bitbucket {
 
     const baseUrl = config.serverUrl || DEFAULT_BITBUCKET_URL;
```
Comment (ypc-faros) [inline sources/bitbucket-source/src/bitbucket.ts:63]: `serverUrl` => `api_url` (you renamed the field in the spec)

File: sources/bitbucket-source/src/bitbucket.ts [hunk 5]
```diff
@@ -104,19 +104,16 @@ export class Bitbucket {
       ];
     }
 
-    if (!config.workspaces || config.workspaces.length < 1) {
-      return [false, 'No workspaces provided'];
-    }
     try {
       config.serverUrl && new URL(config.serverUrl);
```
Comment (ypc-faros) [inline sources/bitbucket-source/src/bitbucket.ts:108]: Same here

File: sources/bitbucket-source/src/streams/common.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  ): StreamState {
+    return calculateUpdatedStreamState(
+      latestRecordCutoff,
+      currentStreamState,
+      workspaceRepoKey
+    );
+  }
+
+  static workspaceKey(workspace: string): string {
+    return toLower(`${workspace}`);
```
Comment (ypc-faros) [inline sources/bitbucket-source/src/streams/common.ts:50]: Why not just `toLower(workspace)` ?

File: sources/bitbucket-source/src/workspace-repo-filter.ts [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+  private readonly filterConfig: FilterConfig;
+  private workspaces?: Set<string>;
+  private reposByWorkspace: Map<string, Map<string, Repository>> = new Map();
+
+  constructor(
+    private readonly config: BitbucketConfig,
+    private readonly logger: AirbyteLogger
+  ) {
+    const {workspaces, repositories, excluded_repositories} = this.config;
+    let {excluded_workspaces} = this.config;
```
Comment (ypc-faros) [inline sources/bitbucket-source/src/workspace-repo-filter.ts:26]: Same for `excluded_workspaces`. Missing from the spec and BitbucketConfig ?

File: sources/bitbucket-source/src/workspace-repo-filter.ts
Comment (ypc-faros) [inline sources/bitbucket-source/src/workspace-repo-filter.ts:1]: We should add tests for this class. Okay to do in another PR

**Review by ypc-faros — COMMENTED** [review]
Nice cleanup 🙇

**Review by ypc-faros — APPROVED** [review]
